core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala

```diff
@@ -95,10 +95,19 @@ private[spark] class PythonWorkerFactory(
   def create(): (PythonWorker, Option[Long]) = {
     if (useDaemon) {
       self.synchronized {
-        if (idleWorkers.nonEmpty) {
+        while (idleWorkers.nonEmpty) {
           val worker = idleWorkers.dequeue()
-          worker.selectionKey.interestOps(SelectionKey.OP_READ | SelectionKey.OP_WRITE)
-          return (worker, daemonWorkers.get(worker))
+          val workerHandle = daemonWorkers(worker)
+          try {
+            worker.selectionKey.interestOps(SelectionKey.OP_READ | SelectionKey.OP_WRITE)
+          } catch {
+            case c: CancelledKeyException => /* pass */
+          }
+          if (workerHandle.isAlive() && worker.selectionKey.isValid()) {
```

Comment:
If the `interestOps` call succeed, will both of these checks be automatically true?

Reply:
It seems that this isn't always the case,  i.e. the workerHandle may already see the process being dead and selectionKey update will happily pass. I also check `isValid` for the off-chance, that we got cancelled after interestOps was set.